src/goto-symex/field_sensitivity.cpp

```diff
@@ -0,0 +1,346 @@
+/*******************************************************************\
+
+Module: Field-sensitive SSA
+
+Author: Michael Tautschnig, mt@eecs.qmul.ac.uk
+
+\*******************************************************************/
+
+#include <util/simplify_expr.h>
+#include <util/std_expr.h>
+#include <util/arith_tools.h>
+
+#include <ansi-c/c_types.h>
+
+#include "symex_target.h"
+#include "goto_symex_state.h"
+
+#include "field_sensitivity.h"
+
+/*******************************************************************\
+
+Function: field_sensitivityt::operator()
+
+  Inputs:
+
+ Outputs:
+
+ Purpose:
+
+\*******************************************************************/
+
+void field_sensitivityt::operator()(
+  const namespacet &ns,
+  exprt &expr,
+  bool write) const
+{
+#if 1
+  if(expr.id()!=ID_address_of)
+    Forall_operands(it, expr)
+      operator()(ns, *it, write);
+
+  if(expr.id()==ID_symbol &&
+     expr.get_bool(ID_C_SSA_symbol) &&
+     !write)
+  {
+    ssa_exprt ssa=to_ssa_expr(expr);
+    get_fields(ns, ssa, expr);
+  }
+  else if(!write &&
+          expr.id()==ID_member &&
+          to_member_expr(expr).struct_op().id()==ID_struct)
+  {
+    simplify(expr, ns);
+  }
+  else if(!write &&
+          expr.id()==ID_index &&
+          to_index_expr(expr).array().id()==ID_array)
+  {
+    simplify(expr, ns);
+  }
+  else if(write && expr.id()==ID_member)
+  {
+    member_exprt &member=to_member_expr(expr);
+
+    if(member.struct_op().id()==ID_symbol &&
+       member.struct_op().get_bool(ID_C_SSA_symbol) &&
+       ns.follow(member.struct_op().type()).id()==ID_struct)
+    {
+      ssa_exprt tmp=to_ssa_expr(member.struct_op());
+
+      member.struct_op()=tmp.get_original_expr();
+
+      tmp.type()=member.type();
+      tmp.set(ID_expression, member);
+      tmp.update_identifier();
+
+      expr.swap(tmp);
+    }
+  }
+  else if(write && expr.id()==ID_index)
+  {
+    index_exprt &index=to_index_expr(expr);
+    simplify(index.index(), ns);
+
+    if(index.array().id()==ID_symbol &&
+       index.array().get_bool(ID_C_SSA_symbol) &&
+       ns.follow(index.array().type()).id()==ID_array &&
+       index.index().id()==ID_constant &&
+       to_array_type(ns.follow(index.array().type())).size().id()==ID_constant &&
+       !to_array_type(ns.follow(index.array().type())).size().is_zero())
+    {
+      ssa_exprt tmp=to_ssa_expr(index.array());
```

Comment:
Comment somewhere about how constant and non-constant array-index operations interact